addons/format_lame.c

```diff
@@ -0,0 +1,541 @@
+			return -2;
+		}
+		if (headers.samplerate != sample_rate) {
+			ast_log(LOG_ERROR, "LAME decoder invalid sampling rate: %d, expected %d", headers.samplerate, sample_rate);
+			return -3;
+		}
+	}
+
+	return samples;
+}
+
+static int mp3lame_file_read(struct ast_filestream *s, int sample_rate)
+{
+	int rres;
+	struct mp3lame_private *p = s->_private;
+	unsigned char dinput[p->total_bytes_compressed];
+
+	rres = fread(dinput, 1, p->total_bytes_compressed, s->f);
+	if(rres != p->total_bytes_compressed){
+		ast_log(LOG_ERROR, "Short read (%d/%d)", rres, p->total_bytes_compressed);
+		return -1;
+	}
+
+	ast_debug(3, "LAME decoder input: %d bytes", rres);
+	p->dsamples = mp3lame_hip_decode(p->hgfp, rres, dinput, p->doutput, sample_rate);
+	while (p->dsamples == 0) {
+		p->dsamples = mp3lame_hip_decode(p->hgfp, 0, NULL, p->doutput, sample_rate);
+	}
+	ast_debug(3, "LAME decoder got %d output samples", p->dsamples);
+
+	return p->dsamples * BITS;
+}
+
+static int mp3lame_open(struct ast_filestream *s)
+{
+	struct mp3lame_private *p = s->_private;
+
+	p->encoder_counter = 0;
+	p->decoder_counter = 0;
+
+	return 0;
+}
+
+static void mp3lame_close(struct ast_filestream *s)
+{
+	struct mp3lame_private *p = s->_private;
+
+	if (p->encoder_counter != 0) {
+		mp3lame_encoder_deinit(s);
+	}
+
+	return;
+}
+
+static struct ast_frame *mp3lame_read(struct ast_filestream *s, int *whennext, int frame_size, int sample_rate)
+{
+	int remaining_samples;
+	struct mp3lame_private *p = s->_private;
+
+	if (p->encoder_counter != 0) {
+		ast_log(LOG_ERROR, "This filestream is already in encoder mode");
+		return NULL;
+	}
+
+	if (p->decoder_counter == 0) {
+		p->total_bytes_compressed = mp3lame_decoder_init(s);
+		p->total_bytes_decoded = mp3lame_file_read(s, sample_rate);
+		p->doffset = mp3lame_decoder_deinit(p);
+	}
+	p->decoder_counter++;
+
+	AST_FRAME_SET_BUFFER(&s->fr, s->buf, AST_FRIENDLY_OFFSET, frame_size);
+
+	remaining_samples = p->dsamples - (frame_size / BITS);
+	if (remaining_samples >= 0) {
+		/* Output a full ast_frame */
+		s->fr.datalen = frame_size;
+		s->fr.samples = frame_size / BITS;
+		*whennext = frame_size / BITS;
+		memcpy(s->fr.data.ptr, (char *)p->doutput + p->doffset, frame_size);
+		p->doffset += frame_size;
+		p->dsamples = remaining_samples;
+		return &s->fr;
+	} else {
+		/* Output a partial ast_frame */
+		if (p->dsamples > 0) {
+			s->fr.datalen = p->dsamples * BITS;
+			s->fr.samples = p->dsamples;
+			memcpy(s->fr.data.ptr, (char *)p->doutput + p->doffset, p->dsamples * BITS);
+			*whennext = 0;
+			p->dsamples = 0;
+			return &s->fr;
+		} else {
+			p->dsamples = 0;
+			return NULL;
+		}
+	}
+}
+
+static int mp3lame_write(struct ast_filestream *fs, struct ast_frame *f, int sample_rate)
+{
+	int res, wres;
+	unsigned char encoder_buffer[sample_rate];
+	struct mp3lame_private *p = fs->_private;
+
+	if (p->decoder_counter != 0) {
+		ast_log(LOG_ERROR, "This filestream is already in decoder mode");
```

Comment:
This error message should be more user friendly

Reply:
I've printed the ast_filestream struct, not sure what else these can say.